src/mito2/src/memtable/time_series.rs

```diff
@@ -0,0 +1,840 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::btree_map::Entry;
+use std::collections::{BTreeMap, Bound};
+use std::fmt::{Debug, Formatter};
+use std::sync::{Arc, RwLock};
+
+use api::v1::OpType;
+use datatypes::arrow;
+use datatypes::arrow::row::RowConverter;
+use datatypes::data_type::DataType;
+use datatypes::prelude::{MutableVector, ScalarVectorBuilder, Vector, VectorRef};
+use datatypes::value::ValueRef;
+use datatypes::vectors::{
+    Helper, UInt64Vector, UInt64VectorBuilder, UInt8Vector, UInt8VectorBuilder,
+};
+use snafu::{ensure, ResultExt};
+use store_api::metadata::RegionMetadataRef;
+use store_api::storage::ScanRequest;
+
+use crate::error::{
+    CompactValuesSnafu, ConvertVectorSnafu, InvalidMemtableConfigSnafu,
+    PrimaryKeyLengthMismatchSnafu, Result, SortValuesSnafu,
+};
+use crate::memtable::{BoxedBatchIterator, KeyValues, Memtable, MemtableId};
+use crate::read::{Batch, BatchBuilder, BatchColumn};
+use crate::row_converter::{McmpRowCodec, RowCodec, SortField};
+
+/// Config for [TimeSeriesMemtable].
+#[derive(Debug)]
+pub struct TimeSeriesMemtableConfig {
+    /// The bucket num for memtable.
+    pub bucket_num: usize,
+    /// The initial capacity for each [ValueBuilder] in series.
+    pub initial_builder_capacity: usize,
+}
+
+impl TimeSeriesMemtableConfig {
+    fn validate(&self) -> Result<()> {
+        ensure!(
+            self.bucket_num < 256,
+            InvalidMemtableConfigSnafu {
+                config: format!("{:?}", self),
+            }
+        );
+        Ok(())
+    }
+}
+
+impl Default for TimeSeriesMemtableConfig {
+    fn default() -> Self {
+        Self {
+            bucket_num: 16,
+            initial_builder_capacity: 128,
```

Comment:
`128` may be too large. For an int64, 128 elements require at least 1k bytes. Consider smaller values like 8 and 16.